### ISSUE TITLE
Faster first element of LINQ methods

### ISSUE THREAD

**Issue by varocarbas** [dotnet/corefx#14561]
I have deleted these contents because of being (partially) faulty and misleading regarding the performance improvements which might be achieved by relying on the proposed methodology.

**Comment by antiufo** [comment]
Another option is simply using `first.First(x => second.Contains(x))`.

**Comment by varocarbas** [comment]
@antiufo Thanks for the proposal but it doesn't seem a good idea as far it performs much worse.

I hope that you don't mind me using your comment to highlight that the previous issue got "out of control" too quickly. This is a quite wide situation with lots of opinions and its seems to trigger certain kind of behaviour which don't think that are too positive for everyone.

Please, before suggesting anything do at least some quick tests with the proposed code. Otherwise, this thread might quickly become a busy chat room not outputting anything worthy.

**Comment by antiufo** [comment]
These are very specific optimizations that are better left to individual developers.

If the second collection contains a lot of duplicate items, it can make sense to turn it first into an hashset. If you need all the elements of the intersection, the current implementation is fine. If you expect the search to finish very soon, it can make sense to use a custom approach like the one above, or a `foreach` or `for` if you also want to avoid allocations.

Nothing is stopping you from writing a custom extension method that performs exceptionally well in your specific scenario. We've all written custom optimizations in our programs. This doesn't mean we should bloat our standard libraries with all conceivable optimizations and utility methods.

**Comment by varocarbas** [comment]
@antiufo

> These are very specific optimizations that are better left to individual developers.

I am proposing a general approach (I am very specific to avoid worthless abstract discussions) which, under the current conditions, performs much better than the standard LINQ methods for almost any size (having two collections of 100 elements isn't too weird and the proposed quick approach is already much quicker there), why individual developers should care about that? You can make the in-built methods to deliver a much better performance under certain (but pretty wide though) conditions and you decide to not do so?

> you need all the elements of the intersection, the current implementation is fine. If you expect the search to finish very soon,

You seem to have missed my point here. I am proposing an approach allowing the existing LINQ methods to perform notably better. It isn't about `Intersect()`, but about all of them; `Intersect()` is just the first one I am analysing. The reason for choosing `First()` is enabling what makes my proposal to perform better than the standard implementations: I am ignoring all the part which is later used while iterating through the `IEnumerable` output; the problem is that I need something (different than an `IEnumerable`) and there is where `First()` came (although in the future, further situations will also be considered). The idea is having a class with optimised versions of all the LINQ methods delivering the first item much quicker.

> This doesn't mean we should bloat our standard libraries with all conceivable optimizations and utility methods.

Bear in mind that we are talking about LINQ methods, which are systematically used by almost every programmer. And also that we are talking about a very relevant performance improvement. Imagine that you can get +7 times faster the result you are looking for (using `First()` after another LINQ call is quite common; I do this a lot myself). Additionally, this issue is mostly about further analysing a very promising line which might eventually drive to a relevant optimisation of all the LINQ methods under virtually any scenario. 

Honestly, I cannot understand your attitude ("No. End of the story") for a so open, promising and still-just-starting line. But it is your opinion and I respect it.

**Comment by varocarbas** [comment]
The code below these lines is expected to address future concerns about the proposed modifications being too specific.

```C#
public static TSource IntersectFirst2<TSource>(IEnumerable<TSource> first, IEnumerable<TSource> second)
{
    if (first == null || second == null) throw new Exception();

    foreach (TSource element in second)
    {
        if (first.Contains(element))
        {
            return element;
        }
    }

    throw new Exception(); //To emulate First() behaviour.
}
```

This is much faster than the original `IntersectFirst`, but it doesn't seem to transmit the intended ideas too well. This code can only deliver the first intersected element, but nothing else. What I am proposing here is a generic approach which is applicable to most of LINQ methods (i.e., ignoring all what is required by future iterations when such scenario isn't happening).

`Intersect` delivers the whole intersection (`IEnumerable` output) of any two collections much quicker than the standard method. The call to `First` in `IntersectFirst` represents the required transition outside `IEnumerable` (i.e., to avoid the performance gain to disappear while iterating through that output) which is being considered in the current issue (without forgetting that different approaches might also be used). On the other hand, the actual implementation will have to be as fast as possible and, consequently, `IntersectFirst2` should always be preferred.

Lastly, I want to highlight that, in principle, I am planning to keep uploading modified versions of all (or most of) the LINQ methods regularly (every few days, I guess).

**Comment by varocarbas** [comment]
Before starting with other methods (the next one in a few days), I want to share my clearer ideas regarding the iterating/just-first-element differences. 

Actually, the problem isn't iterating, but iterating through a big enough number of elements. By taking the aforementioned `Intersect` as reference, there is a trade-off between keeping [the original Set<TSource> set](https://github.com/dotnet/corefx/blob/master/src/System.Linq/src/System/Linq/Intersect.cs) and removing it as proposed. The second option allows a better performance while iterating through the first elements until certain point (that's why `First` doesn't provoke a relevant dropdown).

For `Intersect`, It seems that the new approach performs better up to 3 elements under virtually any scenario (in my computer, the new approach is better for the proposed bigger inputs up to 5 elements). That is, the improved `Intersect` can deliver an always-performing-better `IEnumerable` variable including the first 3 elements. 

In any case, it seems that other methods will have different conditions (+ this issue was created to output the first element), so I will better stick to my original `First` intention.

**Comment by varocarbas** [comment]
I have deleted my comments after realising that might have been too focused on trying to maximise the observed differences without having firstly created a proper algorithm. The proposed `Intersect` algorithm was incomplete and, what is much worse, the performance of the new version was too variable on account of the size of the collection being considered.

This whole evolution (the current issue, the previous one and the original PR) has been quite bad and defined by the short-sightness of my analysis (or better: being much more interested in finding something relevant quickly than in properly analysing the situation) and do apologise (again) about my behaviour. 

In any case, I can confirm that have seen some differences (not beyond 10-20% faster and without having done many tests) in the performances of various LINQ methods when trying a more "short-term" approach (i.e., meant to only iterate through the few first elements) as proposed. Although at this point, I don't think that this is a worthy conclusion to continue this whole analysis: getting the first element +7 times is promising, but not just a bit quicker.

@karelz Perhaps you should close this issue.

**Comment by jnm2** [comment]
@varocarbas You can close it too since you're the author:
![Close button](http://i.imgur.com/woKOEOa.png)

**Comment by varocarbas** [comment]
@jnm2 Thanks for the comment (this time your contribution has been helpful :)). Kind of new in the GitHub contribution world. Additionally, closing an issue has usually been a .NET Team decision, so no need to use that information until now.

**Comment by jnm2** [comment]
@varocarbas You're funny :'D

**Comment by varocarbas** [comment]
@jnm2 In fact and modesty apart, I am quite funny although didn't try to be funny in my last comment (that's why there you have your confused emoji back :)). 

Just to avoid any misunderstanding, bear in mind that your net contribution to the previous issue was none (or even perhaps negative). I do recognise that it was my fault because of not having written clear enough ideas regarding what all that was about, but this doesn't make your contribution positive (because it wasn't). I can explain you further what all this was about (since the start until the moment I decided to abandoned it and how your good but irrelevant comment contributed) if you wish.

**Comment by jnm2** [comment]
Since we're being blunt with each other, what I find funny is that people don't usually devote this much time to analyzing and sharing the details of exactly how justified everyone's actions were, or how we're actually really funny. And of course I'm not entirely able to agree with your perspective, not that it matters. This is all in good faith, my friend. =)

I'm quite happy with my "even perhaps negative" and "good but irrelevant comments". Peace. =)

**Comment by varocarbas** [comment]
Perhaps having deleted potentially-misleading might be negative for future readers to adequately understand what all this was about. I want to maximise my (and others') efforts and avoid people getting unnecessarily confused, so here goes a short summary:
* I submitted a PR with a modified version of `Intersect` performing notably better. As proven there, my tests were incomplete because of only considering the return scenario (`var result = collection1.Intersect(collection2);`), but not the iterating one (`foreach(var item in collection1.Intersect(collection2)){ }`) where it was performing much worse. Despite being a (partially) faulty algorithm, the notable differences returning/iterating seemed very relevant and further-analysing-worthy to me.
* I opened a new issue to analyse this whole situation. That issue was too abstract and imprecise and didn't output anything worthy. Its goal was coming up with the best way to extract the information in the (quicker-generated) `IEnumerable` variable without iterating through it (or creating an alternative iterating methodology or, at least, understanding the reasons for the observed differences).
* Finally, I opened this second issue to focused on `First()` (i.e., getting the `IEnumerable` variable by relying on the new methodology, but quickly converting it into an individual one via `First`). The starting point was seeing a very relevant difference in `Intersect()` (getting the resulting first element up to +7 times faster), what seemed a relevant enough improvement to support a relevant extension (i.e., a new copy of all or most of the LINQ methods to only get the first element). 
I tested different methods and saw noticeably differences, although much smaller (about 10-20% faster depending upon the method and conditions) than originally thought. When I analysed the proposed new `Intersect`, I realised that it wasn't accounting for all the possible scenarios. Bear in mind that I was aware about it not being perfect, although didn't expect that erroneous part to take care of the big difference (+ I wasn't able to run the CoreFX tests on my Windows 7 computer due to a CoreFX bug, which BTW has already been fixed). 
In any case, note that I didn't analyse this situation (i.e., notable differences in the performance of getting and iterating through a Enumerable variable with different approaches) in too much deep. I was looking for a very relevant difference which was proven nonexistent, but this is curious situation which might be optimised at some point (although a quite relevant work is likely to be involved).

@jnm2 Your confusing comment has been a good motivation to write this summary. Hopefully, it is clear enough, sorry for not having added any explanatory pic (-> see? I told you :)).

**Comment by varocarbas** [comment]
>  people don't usually devote this much time to analyzing and sharing the details of exactly how justified everyone's actions were

Actually, I never had that behaviour before but, after living surrounded by misunderstanding-prone people for a while, certainly have this tendency now. Horrible. Anyway, the summary above will hopefully compensate for all these off-topic ideas.

**Comment by jnm2** [comment]
I'm content to be considered wrong unless it is somehow to another person's benefit for me to set it straight.

**Comment by varocarbas** [comment]
Despite the aforementioned summary, I think that this whole issue (and the applicability of its conclusions) is still a bit unclear. That's why I am better including this code.

```C#
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;

namespace LinqQuickFirst
{
    public class Program
    {
        static void Main(string[] args)
        {
            int count = 0;
            int maxCount = 5000000;

            IEnumerable<int> input11 = Enumerable.Range(0, 10000);

            Stopwatch sw = new Stopwatch();
            sw.Start();

            while (count < maxCount)
            {
                count++;

                //var firstAfter10 = input11.Skip(10).First();
                //var firstAfter10Quick = LinqQuick.SkipFirstTen(input11);

                //foreach (var distinctItem in input11.Distinct().Take(10)) { }
                //foreach (var distinctItemQuick in LinqQuick.DistinctTenFirst(input11)) { }
            }

            sw.Stop();

            Console.WriteLine(sw.ElapsedMilliseconds.ToString());
            Console.ReadLine();
        }
    }

    public class LinqQuick
    {
        public static IEnumerable<TResult> DistinctTenFirst<TResult>(IEnumerable<TResult> source)
        {
            if (source == null) throw new Exception();

            List<TResult> done = new List<TResult>();
            int count = 0;
            foreach (var item in source)
            {
                if (!done.Contains(item))
                {
                    if (count == 10) yield break;
                    count++;
                    done.Add(item);
                    yield return item;
                }
            }
        }

        public static TSource SkipFirstTen<TSource>(IEnumerable<TSource> source)
        {
            if (source == null) throw new Exception();

            int count = 0;
            foreach (TSource element in source)
            {
                count++;
                if (count == 11) return element;
            }

            throw new Exception();
        }
    }
}
```

This is a simplified version of the code I firstly wrote for this issue; a simple but descriptive enough testing program. Note that the included methods (modified versions of LINQ `Distinct` and `Skip`) have two features which aren't present in any of the other references so far (i.e., this issue, the previous one and the original PR):
* Their original codes ([Distinct](https://github.com/dotnet/corefx/blob/master/src/System.Linq/src/System/Linq/Distinct.cs) and [Skip](https://github.com/dotnet/corefx/blob/master/src/System.Linq/src/System/Linq/Skip.cs)) have nothing to do with the ones discussed so far (`Intersect` and mentions to the almost-identical `Except`), what supports the general applicability of the proposed ideas (i.e., the `System.Linq` methods optimised to deliver the overall best performance while iterating through all the elements of the corresponding `Enumerable` variable, but not necessarily for just the first elements).
* Both cases analyse up to the 10th element and one of them returns a `IEnumerable` variable (by proving that it is possible to build quicker versions not just for the first element/non-`Enumerable`). 

I am not including the time differences because they are quite variable (on account of the given computer and inputs) and, what is more important, not too big (that's why I closed this issue). Anyone can take this code, modify it, do some tests and confirm the notable differences. Note that, since the first moment, I have tested the time differences as follows: generating the executable for each scenario, running each of them three times and comparing the average time measurements with the equivalent approach (i.e., `firstAfter10` vs. `firstAfter10Quick` and both `foreach`).

At this point, I see one worthy conclusion from all this: if you are planning to use LINQ queries and are only interested in the first elements, perhaps you should do some tests with a custom code rather than blindly relying on the in-built methods.

**Comment by jnm2** [comment]
One thought is that `List` is faster than `Set` for small sizes like 10. Besides that I can't imagine that the LINQ is much different.

**Comment by jnm2** [comment]
* Other than that specialized is always gayer than general-purpose code, but that's a universal principle.

**Comment by varocarbas** [comment]
@jnm2 

> One thought is that List is faster than Set for small sizes like 10. Besides that I can't imagine that the LINQ is much different.

Some ideas for such a piece of... something:
- `Set` isn't used in `Skip` and  the algorithm of the in-built version of `Distinct` differs from the modified one (which is just a simple version I wrote in some minutes after a few tests) by more than just replacing `Set` with `List`.
- I didn't say that 10 was the absolute maximum. It was a descriptive enough scenario to explain the point (-> as previously; you aren't too good at adequately understanding examples/context, are you? Should I have included a picture?). Depending upon sizes, collection types and methods you might go well beyond 10.
- The whole point of all this from the start was to propose a set of improved-under-specific conditions LINQ methods, not modifying the existing ones. Adequately understanding these peculiarities (still not the case, neither interested in getting there) might have allowed to further improve the existing LINQ methods (you know? The better you understand something, the most likely is that you can improve it). Even without doing that, some programmers might apply the suggested conclusions at their work (again, the more information, the better the work; surprising, isn't it?).

Honestly, I don't know what is the exact goal of your comment. Exactly the same than what happened with all your previous ones, except the one with the picture (thanks again!).

**Comment by varocarbas** [comment]
@jnm2 

> Other than that specialized is always gayer than general-purpose code, but that's a universal principle.

Seriously, what is wrong with you? Can you please stop writing off-topic nonsense in an-already-quite-unclear thread?

**Comment by jnm2** [comment]
@varocarbas An obvious phone typo which I immediately corrected.

I would answer your replies last except that given the ridiculously hostile tone, I'm done. Even if what you said about me is true, I still don't deserve to have to put up with this kind of aggressive behavior.

**Comment by varocarbas** [comment]
@jnm2 Seriously, what is wrong with you? (I deleted my previous reference because you clarified that what provoked such a reply was a mistake and didn't any problem. But then you added it to your previous post!?).

FYI, the one you are referring is my first issue here, a couple of people like you started a thread in Reddit behind my back to talk about my behaviour?! (I knew about it the day afterwards, but never got there neither read anything. So pathetic, although apparently you think that this a normal behaviour. Surprise!). Your pointless (since the first second you came here to bother), voting-focused (have you had fun down voting my comments? I upvoted yours, BTW, you should be very proud of that marvellous picture of yours) and asking-for-mods-after-creating-chaos-by-your-own attitude (remember: you self-invited yourself since the first moment, every second. Are you in love with me, or something? Sorry but I don't like you. Be brave and continue looking for someone with your "peculiarities") tells me that you come from StackOverflow or similar (nothing against this site, but lots people like you that's why I don't use it). BTW, I have got a 4 downvotes in a row in random answers of my SO account. Do you know about that? Sure you know! (pathetic). You should know that things here work differently (you have much more votes in your repositories than me! You knew about the button for closing an issue! How can you not know about this?). But you can call a .NET Team member to tell them about your problems (not sure if having to hear your nonsense is part of their work; you know? start crying and replace "daddy", "teacher" or similar with ".NET member"). For example, the one assigning the tags to this issue @karelz 

FYI, the "mods" as you call them, asked me for reducing my aggressive tone and I have been doing it since then (accepting that not everyone has the same knowledge, expectations, objectives, that there are spoiled pathetic kids, etc. Better avoiding problems and ignore some people). But I don't think that anyone reading your comments since the start (in the other thread) will defend you, what you brought (1 picture + random up/down votes + random emojis + offtopic nonsense) and what you represent. If that person exist, I would think of him/her exactly the same than of my Reddit fans, the SO random downvoters and you.

Bye cry-baby, I will not spend a single second of my life with a "person" like you. Go up/down-vote something and don't bother me.

**Comment by jnm2** [comment]
I rest my case.
